[크리스마스 프로모션] 김세민 미션 제출합니다.

seminss opened the pull request.

seminss committed:
docs: 기능 명세서 작성

seminss committed:
feat: 예상 방문 날짜 입력을 검증한다.

seminss committed:
feat: 주문 메뉴와 개수 입력을 검증한다.

seminss committed:
feat: 예상 방문 날짜를 입력 받는다.

애플리케이션의 뼈대 코드 생성

seminss committed:
feat: 메뉴판을 준비한다.

seminss committed:
feat: 주문 메뉴와 개수를 입력 받는다.

주문 메뉴에 대한 비즈니스 조건을 모델에서 검증한다.

seminss committed:
test: 주문 메뉴와 개수에 대한 비즈니스 조건 예외를 테스트한다.

seminss committed:
feat: 선택된 메뉴를 바탕으로, 할인 전, 총 주문 금액을 연산한다.

seminss committed:
feat: 총 주문 금액에 따른 이벤트 진행 여부를 확인한다.

seminss committed:
feat: 총 주문 금액에 따른 샴페인 증정 여부를 확인한다.

seminss committed:
feat: EventCalendar 구현, 할인 정책이 적용 되는 날인지 확인한다.

seminss committed:
feat: 크리스마스 디데이 할인 금액을 연산한다.

seminss committed:
refactor: visitDate의 자료형을 LocalDate로 수정

seminss committed:
refactor: InitialOrderAmount을 Order 내부에 저장, 계산 로직 추가

seminss committed:
refactor: 특별한 날, 증정품 별도 클래스로 분리

확장성 및 재활용 위함

seminss committed:
feat: 할인 정책을 위한 ChristmasDiscountCalulator가 할인을 적용한다.

연산을 위한 모델 생성.
DiscountType: 할인 정책 종류
DiscountAmount: 각 정책별 할인 금액
DiscountResult: Type과 Amount 매핑, 총합 연산
ChristmasDiscountPolicys: 크리스마스 할인 정책 통해 연산
ChristmasDiscountCalulator: 각 할인을 종합해 DiscountResult 생성

seminss committed:
refactor: 디렉토리 구조 정리 및 테스트 클래스 명 업데이트

Order에 대한 validation과 util 테스트를 분리하기 위함

seminss committed:
test: 할인 정책 연산을 테스트 한다.

주말을 토요일, 일요일로 계산하고 있던 버그 fix

seminss committed:
refactor: Calculator에서 visitDate, Order를 메서드 파라미터로 넘기도록 수정.

증정 이벤트를 할인 정책에서 분리 했으나, >>> 11/14 다시 병합하고, Giveaway정의는 Enum에서 한다.

seminss committed:
feat: 주문 메뉴와 개수, 총 주문 금액을 출력 한다.

DetailsFormatter: Service에서 생성된 결과를 Formatting 해주는 객체

seminss committed:
feat: 할인 후 예상 결제 금액과, 이벤트 배지를 결정한다.

seminss committed:
feat: 우테코 식당에서 받을 이벤트 혜택을 출력한다.

Giveaway: 증정품을 관리하는 enum class
PromotionSummary: 이벤트 혜택을 출력하기 위해 view에 넘기는 결과
DiscountResult: 혜택 종합 결과(총할인 금액, 최종 결제 금액, 배지)

seminss committed:
fix: 배지를 결정할 때 할인 혜택을 통해 결정 하도록 수정

예상 결제 금액으로 결정 하고 있었던 버그.

seminss committed:
fix: 할인 혜택이 없는지 체크 하는 로직 수정

모든 할인 정책이 적용 되지 않고 있는지 제대로 판단 하지 못하던 버그.
단순 개행(\n) 이 아니라 System.lineSeparator() 호출 해서 개행

seminss committed:
refactor: 디렉토리 구조 개선 및 네이밍 통일

DiscountBenefits -> DiscountItems
DiscountConfig -> DiscountSettings
DiscountResult -> OrderDiscountDetails
PromotionSummary 의 메서드 명 수정
Order의 InitialOrderAmount 필드 제거 (이유: 메서드로 반환 가능)

seminss committed:
feat: ChristmasPromotion에서 input 에러 핸들링을 한다.

에러 메세지 정리: view와 model에서 전부 IVALID_ORDER, INVALID_DATE 만 사용한다. (음료만 주문하는 경우 제외)

seminss committed:
refactor: 계층 구조 정리

ChristmasPromotion-> controller로 수정
기존 controller와 service를 병합

seminss committed:
refactor: input 에러 핸들링 로직을 InputView로 이동

supplier을 사용해 함수로 valid input을 반환하도록 설정

seminss committed:
refactor: PromotionSummary를 record-> class 로 변경

10000원을 넘지 못하면 Default 생성자 반환 하기 위함. (이벤트 적용 x)
변수 및 메서드 네이밍 수정

seminss committed:
feat: 총 주문 금액이 10000원을 넘지 않으면 혜택 없이 미리보기를 제공한다.

seminss committed:
refactor: order과 visitDate에 대해서도 Summary를 생성 후 출력

DetailsFormatter의 간소화를 위함.

seminss committed:
refactor: 이벤트 캘린더 Java.time 사용하여 하드코딩 제거

seminss committed:
test: 특별 할인일 초기화 테스트

ChristmasSpecialDays -> ChristmasSpecialDayEvaluator로 클래스명 변경

seminss committed:
test: 할인 정책 적용 판단 테스트

ChristmasEventCalendar ->ChristmasEventSchedular로 클래스명 변경
주말, 평일, 증정, 특별, D-Day 할인일 핸들링 테스트

seminss committed:
test: 할인 정책 별로 연산이 올바르게 되는지 테스트

seminss committed:
test: ChristmasDiscountCalculator 서비스 테스트

seminss committed:
refactor: 뱃지 결정 로직 정적 팩토리 메서드로 추출

seminss committed:
fix: 증정품 갯수를 가격으로 출력하고 있던 버그 수정

seminss committed:
test: 총혜택 금액 및 할인 후 예상 결제 금액 연산 테스트

seminss committed:
fix: 혜택 금액을 양수, 음수 혼용하여 뱃지 출력 못하던 버그 수정

seminss committed:
docs: 디렉토리 구조 및 예시 추가

테스트코드 디렉토리 구조 조정

seminss committed:
refactor: 불필요한 print 메서드 삭제

printPromotionPreviewMessage, printFormattedDiscountResultMessage, printFormattedOrderMessage 가 전부 같은 역할을 하고 있어, 병합.

seminss committed:
refactor: EventSchedular 클래스명 수정 (구 EventCalendar)

ChristmasEventSchedular의 부모 클래스

seminss committed:
fix: 모델에서 발생하는 예외도 다시 입력 받을 수 있도록 수정

inputView의 getValidInput 메서드(supplier사용)제거, controller에 run 메서드(runnable사용) 추가

seminss committed:
refactor: 빠진 접근 제어자 및 final 선언 추가

chanil4129 commented on src/main/java/christmas/config/Menu.java:
```diff
@@ -0,0 +1,54 @@
+    private final int price;
+    private final Category category;
+
+    public enum Category {
+        APPETIZER, MAIN, DESSERT, DRINK, INVALID
+    }
```
enum 안에 enum 몰랐는데 알아갑니다.

Octoping925 replied:
Category가 private면 더 좋을 것 같아요.

사실상 `category == Menu.Category.DRINK` 이런 용례로만 사용되고 있습니다.
Menu에게 isDrink 냐 와 같은 식으로 알려달라고 시키면 어떨까요?

seminss replied:
그렇네요. 당장 적용시키고 싶은 좋은 의견 감사합니다!

seminss committed:
refactor: 출력 순서는 `크리스마스 디데이` - `평일` - `주말` - `특별` 할인 순이다.

chanil4129 commented on src/main/java/christmas/view/input/parser/OrderParser.java:
```diff
@@ -0,0 +1,20 @@
+import static christmas.constant.EventSymbol.MENU_SEPARATOR;
+
+public class OrderParser {
+    public static List<SimpleEntry<String, Integer>> parseEachMenu(String userInput) {
```
SimpleEntry에 대해 처음 알았는데 자료구조 사용 너무 좋은거 같아요.

chanil4129 replied:
DTO 활용해도 좋을 것 같아요.

seminss replied:
의견 감사합니다 ~:)

Octoping925 replied:
엔트리라는 타입으로는 String과 Integer에 해당되는 값이 무슨 값인지 알기 어렵습니다.
객체로 만들어서 변수명으로 의미가 드러났으면 합니다

( + 그러면 개수는 음수가 될 수 없다던가 한 validation 같은 로직들도 그 객체에서 수행할 수 있겠죠)

seminss replied:
고민이 되었던 부분이긴 한데요, VisitDate, Order라는 모델이 있기 때문에 service 단으로 넘어가서 모델을 정의하기 전까지는 딱히 명시를 해주지 않아도 된다고 생각했던 것 같습니다.
그런데 DTO로 만들어서 넘기면 inputView에서의 로직은 확실히 간단해질 것 같습니다. validator, parser같이 복잡하게 클래스를 나눌 필요가 없어지니까요! service와 controller를 무상태하게 설계하는 데 도움이 되었을까 싶기도 하고요. 리펙토링하면서 적용해 봐야겠습니다! 두 분 의견 감사합니다.

chanil4129 reviewed (commented):
코드 잘 읽었습니다. 구조가 너무 좋은 것 같아요.

develsvai commented on src/main/java/christmas/model/policy/calendar/EventSchedular.java:
```diff
@@ -0,0 +1,13 @@
+    public abstract boolean isSpecialDiscountDay(LocalDate date);
+
+    public abstract boolean isDDayDiscountDay(LocalDate date);
+}
```
이벤트 스케줄러 를 추상 클래스를 사용한 이유가 궁금해요!

develsvai commented on src/main/java/christmas/model/Order.java:
```diff
@@ -0,0 +1,104 @@
+            orderedMenu.put(menu, quantity);
+        }
+        validateTotalQuantity(orderedMenu.values());
+    }
```
for문을 메소드 로 한번더 분리 해도 좋을 것 같아요 !

mingeun2154 commented on src/main/java/christmas/controller/ChristmasPromotionController.java:
```diff
@@ -0,0 +1,70 @@
+        return DetailsFormatter.DiscountResultFormatter(service.getPromotionSummary());
+    }
+
+    private void run(Runnable inputRunnable) {
+        while (true) {
+            try {
+                inputRunnable.run();
+                break;
+            } catch (IllegalArgumentException e) {
+                OutputView.printMessage(e.getMessage());
+            }
+        }
+    }
```
반복되는 동작을 이렇게 메서드로 분리할 수 있다는점 배워갑니다.
다만 메서드 이름을 run보다는 runUntilValidInputDetected처럼 의도가 드러나도록 만드는게 다른 사람과 함께 작업할때 의도를 분명하게 전달할 수 있을것 같아요.

seminss replied:
의견 감사합니다! 함수형 인터페이스 관련해서 검색해보시면 좋을 것 같아요 😁

develsvai replied:
덕분에 람다와 함수형 인터페이스 에 대해 많이 알게 되었습니다! 배워갑니다!

mingeun2154 commented on src/main/java/christmas/view/output/OutputView.java:
```diff
@@ -0,0 +1,26 @@
+
+    public static void printMessage(DetailsFormatter detailsFormatter) {
+        printMessage(detailsFormatter.toString());
+    }
```
사용자 UI 를 담당하는 OutputView 클래스로 서비스 객체가 생성한 결과를 DTO에 담아서 전달함으로써 UI와 서비스 객체의 결합도를 낮추셔서 콘솔이 아닌 다른 플랫폼으로 이식이 편할 것 같습니다.

mingeun2154 commented on src/main/java/christmas/service/ChristmasDiscountCalculator.java:
```diff
@@ -0,0 +1,85 @@
+        return new DiscountedItems(discounts);
+    }
+
+    private boolean isWeekday(VisitDate visitDate) {
+        return eventSchedular.isWeekday(visitDate.getDate());
+    }
+
+    private DiscountAmount discountSpecialDay(VisitDate visitDate) {
+        if (eventSchedular.isSpecialDiscountDay(visitDate.getDate())) {
+            int amount = discountPolicy.calculateSpecialDiscountPrice();
+            return new DiscountAmount(SPECIAL_DISCOUNT, amount);
+        }
+        return new DiscountAmount(SPECIAL_DISCOUNT, 0);
+    }
+
+    private DiscountAmount discountWeekend(VisitDate visitDate, Order order) {
+        if (eventSchedular.isWeekend(visitDate.getDate())) {
+            int mainQuantity = order.getMainQuantity();
+            int amount = discountPolicy.calculateWeekendDiscountPrice(mainQuantity);
+            return new DiscountAmount(WEEKEND_DISCOUNT, amount);
+        }
+        return new DiscountAmount(WEEKEND_DISCOUNT, 0);
+    }
+
+    private DiscountAmount discountWeekDay(VisitDate visitDate, Order order) {
+        if (isWeekday(visitDate)) {
+            int dessertQuantity = order.getDessertQuantity();
+            int amount = discountPolicy.calculateWeekdayDiscountPrice(dessertQuantity);
+            return new DiscountAmount(WEEKDAY_DISCOUNT, amount);
+        }
+        return new DiscountAmount(WEEKDAY_DISCOUNT, 0);
+    }
+
+    private DiscountAmount discountDDay(VisitDate visitDate) {
+        if (eventSchedular.isDDayDiscountDay(visitDate.getDate())) {
+            int amount = discountPolicy.calculateDDayDiscountPrice(visitDate.getDate().getDayOfMonth());
+            return new DiscountAmount(DDAY_DISCOUNT, amount);
+        }
+        return new DiscountAmount(DDAY_DISCOUNT, 0);
+    }
```
서로 다른 할인 정책의 할인 금액을 계산하는 로직은 각각의 할인 정책을 클래스로 만들어서 해당 클래스의 메서드로 정의하는 게 객체의 책임을 작게 만들어 생산성을 높일것 같습니다.
지금의 ChristmasDiscountCalculator는 너무 많은 일을 하고 있다고 생각합니다.

seminss replied:
의견 감사합니다. 재구현 할 때는 조언해주신 방법을 고려해봐야겠네요👍

mingeun2154 commented on src/main/java/christmas/model/Order.java:
```diff
@@ -0,0 +1,104 @@
+        int totalQuantity = values.stream()
+                .mapToInt(Integer::intValue)
+                .sum();
+        if (totalQuantity > 20) {
```
상수는 리터럴을 사용하기보다 static final 변수를 사용하는게 의도를 명확하게 나타내고 리팩토링하기 편합니다.

seminss replied:
하드 코딩된 부분을 놓쳐버렸네요..🥹 짚어주셔서 감사합니다!

mingeun2154 commented on src/main/java/christmas/model/policy/calendar/ChristmasEventSchedular.java:
```diff
@@ -0,0 +1,35 @@
+    private final LocalDate christmasDate = LocalDate.of(now().getYear(), Month.DECEMBER, 25);
+    private final ChristmasSpecialDayEvaluator specialDates = new ChristmasSpecialDayEvaluator();
+
+    @Override
+    public boolean isWeekday(LocalDate date) {
+        DayOfWeek day = date.getDayOfWeek();
+        return day != DayOfWeek.SATURDAY && day != DayOfWeek.FRIDAY;
+    }
```
return !isWeekend(date); 로 변경하면 훨씬 직관적이고 오류가 발생할 가능성도 줄어듭니다.

seminss replied:
평일과 주말을 굳이 따로 구현할 필요가 없었네요..😅 감사합니다ㅎㅎ

mingeun2154 commented on src/main/java/christmas/model/policy/calendar/EventSchedular.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.time.LocalDate;
+
+public abstract class EventSchedular {
+    public abstract boolean isWeekday(LocalDate date);
+
+    public abstract boolean isWeekend(LocalDate date);
+
+    public abstract boolean isSpecialDiscountDay(LocalDate date);
+
+    public abstract boolean isDDayDiscountDay(LocalDate date);
```
이 메서드들도 각각의 이벤트를 표현하는 클래스의 메서드로 만드는 것이 좋다고 생각합니다. 
할인 정책에 대한 코드가 EventSchedular, ChristmasDiscountCalculator 등 여러 클래스에 흩어져있어서 하나의 이벤트 정책을 수정할 경우 여러 클래스를 수정해야 하기 때문입니다.

seminss replied:
그렇네요. 민근님 말씀처럼 정책 별로 클래스를 구분하고, 흩어져있는 메서드들을 다시 적절하게 배분을 해줬다면 전반적으로 역할과 책임이 더 명확해졌을 것 같네요! 조언 감사합니다!

mingeun2154 reviewed (commented):
제가 관심있는 할인 정책 표현, 입출력 과정 위주로 리뷰했습니다.
저보다 잘 하신 점도 있었고 개선이 필요해보이는 부분도 있었습니다.
리뷰를 통해 저의 코드도 객관적으로 돌아볼 수 있었습니다.

Octoping925 commented on src/main/java/christmas/config/Badge.java:
Badge, Giveaway, Menu가 Config에 있는 이유는 무엇인가요? model이 아닌가 싶네요

seminss replied:
프로모션 이벤트를 정의하는 핵심 구성 요소라고 생각해서 config에 넣기는 했지만, model에 넣어도 좋았을 것 같아요.ㅎㅎ 지금 보니 config가 설정 정보라는 느낌이 강하게 드는 것 같기도 하고요,, 고민 되네요 😅

Octoping925 commented on src/main/java/christmas/config/Badge.java:
```diff
@@ -0,0 +1,32 @@
+        return name;
+    }
+
+    public int getThreshold() {
+        return threshold;
+    }
+
+    public static Badge of(int totalDiscountAmount) {
+        for (Badge badge : Badge.values()) {
+            if (Math.negateExact(totalDiscountAmount) >= badge.getThreshold()) {
+                return badge;
+            }
+        }
+        return null;
+    }
```
1. getThreshold가 of에서만 쓰여서, public일 필요 없을 것 같아요! 사실 이렇게 써도 되니 아예 필요가 없기도 하겠네요
```java
if (Math.negateExact(totalDiscountAmount) >= badge.threshold)
```
2. of의 로직은 Badge의 순서에 강하게 의존하고 있어요. enum에는 순서가 없는데, 이 순서를 누군가 바꾸게 된다면 로직이 잘못 실행될 위험이 있어요!

seminss replied:
와 2번 부분은 미처 생각하지 못했네요! 짚어주셔서 감사합니다.

Octoping925 commented on src/main/java/christmas/controller/ChristmasPromotionController.java:
```diff
@@ -0,0 +1,70 @@
+public class ChristmasPromotionController {
+    private final ChristmasPromotionService service;
+
+    public ChristmasPromotionController() {
+        service = new ChristmasPromotionService();
+    }
```
생성자의 파라미터로 주입받는, 생성자 주입을 하면 더 유연한 코드를 작성할 수 있을 것 같아요

seminss replied:
ChristmasPromotion에 관련된 Controller, Service이기 때문에 직접주입 해주어도 상관 없겠다 싶었습니다.
다시 생각해보니 같은 크리스마스 프로모션이여도 service 단이 변경될 수 있겠네요. 의견 감사합니다!

Octoping925 commented on src/main/java/christmas/controller/ChristmasPromotionController.java:
```diff
@@ -0,0 +1,70 @@
+        service = new ChristmasPromotionService();
+    }
+
+    public void start() {
+        takeVisitDate();
+        takeOrder();
+        previewEventBenefits();
+    }
+
+    private void takeVisitDate() {
+        run(() -> {
+            OutputView.printIntroductionMessage();
+            OutputView.printTakeDateMessage();
+            Integer visitDate = InputView.readVisitDate();
+            service.setVisitDate(visitDate);
+        });
+    }
```
서비스는 무상태하게 설계하는게 좋습니다
```java
    public void start() {
        int visitDate = takeVisitDate();
        takeOrder(visitDate);
        previewEventBenefits(visitDate);
    }

    private void takeVisitDate() {
        return run(() -> {
            OutputView.printIntroductionMessage();
            OutputView.printTakeDateMessage();
            int visitDate = InputView.readVisitDate();
            return visitDate;
        });
    }
```
이런 느낌으로 될 수 있었다면? 싶네요

seminss replied:
서비스를 무상태하게 설계해야 한다는 것에는 동의합니다!
그러나 Octoping님 말씀대로 하게 되면 controller가 상태를 가지게 되는게 아닐까 싶습니다.
controller, service 모두 무상태하도록 설계하는 방법을 고민해봐야겠네요. 좋은 의견 감사합니다!

Octoping925 commented on src/main/java/christmas/controller/ChristmasPromotionController.java:
```diff
@@ -0,0 +1,70 @@
+        while (true) {
+            try {
+                inputRunnable.run();
+                break;
```
취향차이긴 한데, 그냥 return 해도 좋을듯요

Octoping925 commented on src/main/java/christmas/view/input/parser/OrderParser.java:
```diff
@@ -0,0 +1,20 @@
+                    String[] parts = item.trim().split(MENU_CONNECTOR.getValue());
+                    return new SimpleEntry<>(parts[0],Integer.parseInt(parts[1]));
+                })
+                .collect(Collectors.toList());
```
toList()를 활용해보면 좋을 것 같아요.
1. 불변성
2. 가독성

Octoping925 commented on src/main/java/christmas/model/summary/PromotionSummary.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    private void initDiscountDetails(List<DiscountAmount> discounts) {
+        if (discounts.stream().allMatch(d -> d.amount() == NO_DISCOUNT_THRESHOLD)) {
+            this.discountDetails = DEFAULT_MESSAGE;
+            return;
+        }
+        this.discountDetails = discounts.stream()
+                .filter(d -> d.amount() < NO_DISCOUNT_THRESHOLD)
+                .map(this::formatDiscountDetail)
+                .collect(Collectors.joining(System.lineSeparator()));
```
1. if 안의 조건문
2. discountDetail을 만드는 과정
각각 private 메소드로 추출해서 의미가 잘 드러나는 이름을 만들 수 있을 것 같아요

Octoping925 replied:
DiscountAmount가 NO_DISCOUNT_THRESHOLD과 같은 값인지 체크하는 로직은, DiscountAmount안에서 이뤄질 수 있을 것 같아요 (Tell, Don't Ask)

seminss replied:
디테일 짚어주셔서 감사합니다👍👍

Octoping925 commented on src/main/java/christmas/model/summary/PromotionSummary.java:
```diff
@@ -0,0 +1,107 @@
+        this.eventBadgeDetails = DEFAULT_MESSAGE;
+    }
+
+    public PromotionSummary(List<DiscountAmount> discounts, int totalDiscount, int finalPaymentAmount,
+                            Optional<Badge> eventBadge) {
+        initDiscountDetails(discounts);
+        initFinalPaymentAmountDetails(finalPaymentAmount);
+        initTotalDiscountDetails(totalDiscount);
+        initEventBadgeDetails(eventBadge);
+        initGiveawayItemDetails(discounts);
+    }
```
함수의 파라미터, 클래스의 필드값으로 Optional을 사용하는 것은 안티패턴입니다.

seminss replied:
넵 덕분에 알아갑니다.
```
    public Optional<Badge> getBadge() {
        return Optional.of(badge);
    }
```
리턴 값으로 넘기게 수정해야겠네요! 감사합니다.

Octoping925 commented on src/main/java/christmas/model/summary/PromotionSummary.java:
```diff
@@ -0,0 +1,107 @@
+                .collect(Collectors.joining(System.lineSeparator()));
+    }
+
+    private void initEventBadgeDetails(Optional<Badge> eventBadge) {
+        if (eventBadge.isEmpty()) {
+            this.eventBadgeDetails = DEFAULT_MESSAGE;
+            return;
+        }
+        this.eventBadgeDetails = eventBadge.get().getName();
+    }
```
어쩌다보니 init 시리즈들이 setter처럼 됐는데, 이런 방식이면 더 좋을 것 같아요
그러면 각 멤버변수들도 final로 만들 수 있을 것 같아요

```java
this.eventBadgeDetails = getEventBadgeDetails(eventBadge);

private String getEventBadgeDetails(Badge eventBadge) {
    if(eventBadge == null) {
        return DEFAULT_MESSAGE;
    }

    return eventBadge.getName();
}
```

seminss replied:
오호🧐 이런 방식도 고려를 안 한 것은 아닌데요, 다양한 코드를 많이 보지 못해서(봤더라도 장점을 크게 알지 못해서) setter처럼 사용하는 게 깔끔하다고 생각해 버린 것 같습니다.
get 하도록 메서드를 정의하면 멤버 변수들을 final로 정의할 수 있다는 장점이 있었네요!! 왜 생각을 못 했을까요!! 앞으로는 initialize나 set 키워드는 최대한 자제하고 getter로 가져와서 정의하도록 노력해야겠네요. 감사합니다:)

Octoping925 commented on src/main/java/christmas/model/policy/calendar/EventSchedular.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.time.LocalDate;
+
+public abstract class EventSchedular {
```
인터페이스가 더 어울렸을 것 같습니다

seminss replied:
처음에는 schedular가 아니라 calendar 개념으로 생각해서 1월, 2월 .. 12월이라 생각하고 구현을 하면서 추상 클래스로 설계했었는데요,
리펙토링을 거치면서 schedular로 변경하고 나니 인터페이스가 더 어울리게 되었네요. Octoping님 말씀에 동의합니다. 의견 감사합니다:)

Octoping925 commented on src/test/java/christmas/config/BadgeTest.java:
```diff
@@ -0,0 +1,33 @@
+    @DisplayName("총혜택 금액이 5천 원 이상은 별이다.")
+    @Test
+    void of1() {
+        Assertions.assertEquals(Badge.of(5000), Badge.STAR);
```
왼쪽이 예상 값, 오른쪽이 실제 값이라서 순서가 반대입니다

seminss replied:
앗,, 이것도 컨벤션이 있군요ㅎㅎ 짚어주셔서 감사합니다!

Octoping925 commented on src/main/java/christmas/model/policy/calendar/ChristmasSpecialDayEvaluator.java:
```diff
@@ -0,0 +1,38 @@
+import static java.time.Month.DECEMBER;
+
+public class ChristmasSpecialDayEvaluator {
+    private static final LocalDate christmasDate = LocalDate.of(now().getYear(), Month.DECEMBER, 25);
+    private static final LocalDate startDateOfDecember = LocalDate.of(now().getYear(), Month.DECEMBER, 1);
```
2023년이 요구사항인지라, 24년이 되면 터지게 될 것 같아요!

seminss replied:
넵 안그래도 년도, 월을 묶어서 깔끔하게 관리할 수 있는 방법을 고민하고 있습니다.
고려해서 리펙토링 해보도록 하겠습니다!

Octoping925 commented on src/main/java/christmas/model/policy/calendar/ChristmasSpecialDayEvaluator.java:
```diff
@@ -0,0 +1,38 @@
+        return specialDays.contains(date);
+    }
+
+    private static Set<LocalDate> initializeSpecialDays() {
+        Set<LocalDate> specialDays = new HashSet<>();
+        //12월의 모든 일요일을 스페셜 데이에 추가
+        LocalDate date = startDateOfDecember.with(TemporalAdjusters.firstInMonth(SUNDAY));
+        while (date.getMonth() == DECEMBER) {
+            specialDays.add(date);
+            date = date.with(TemporalAdjusters.next(SUNDAY));
+        }
+        //크리스마스를 스페셜 데이에 추가
+        specialDays.add(christmasDate);
+        return specialDays;
+    }
```
1. static일 필요가 없을 것 같아요
2. 매 일요일 구하는 로직을 private 메소드로 추출할 수 있을 거 같아요

제 나름대로 리팩토링해보면 이럴 듯 합니다
```java
    private Set<LocalDate> initializeSpecialDays() {
        Set<LocalDate> specialDays = new HashSet<>();

        specialDays.addAll(getAllSundayOfDecember());
        specialDays.add(CHRISTMAS_DATE);

        return specialDays;
    }

    private Set<LocalDate> getAllSundayOfDecember() {
        LocalDate firstDayOfDecember = LocalDate.of(2023, 12, 1);
        LocalDate lastDayOfDecember = LocalDate.of(2023, 12, 31);

        return firstDayOfDecember.datesUntil(lastDayOfDecember.plusDays(1))
                .filter(date -> date.getDayOfWeek() == SUNDAY)
                .collect(Collectors.toUnmodifiableSet());
    }
```

seminss replied:
오 좋은 코드 공유 감사합니다!!! 개인적으로 calendar 디렉토리에 있는 클래스들을 신경을 못써서, 리펙토링 1순위라고 생각하고 있었습니다.
static은 intellij 자동 메서드 추출 기능을 이용하다보니 static으로 추출이 되어버렸는데, 또 놓쳐버렸네요.. 짚어주셔서 감사합니다.

Octoping925 commented on src/main/java/christmas/service/ChristmasPromotionService.java:
```diff
@@ -0,0 +1,50 @@
+
+    public PromotionSummary getPromotionSummary() {
+        if (canReceivePromotion()) {
+            DiscountedItems discountedItems = discountCalculator.calculateDiscounts(visitDate, order);
+            DiscountResults discountResults = new DiscountResults(order.getBaseOrderAmount(), discountedItems);
+            return new PromotionSummary(discountedItems.items(), discountResults.getTotalDiscountAmount(),
+                    discountResults.getFinalPaymentAmount(), Optional.ofNullable(discountResults.getBadge()));
```
이 부분 로직이 길어서 private로 한번 추출할 수 있었을 것 같아요

Octoping925 replied:
early-return을 사용해보면 좀 더 간단해질 것 같아요
```java
public PromotionSummary getPromotionSummary() {
    if (!canReceivePromotion()) {
        return new PromotionSummary(order.getBaseOrderAmount());
    }

    DiscountedItems discountedItems = discountCalculator.calculateDiscounts(visitDate, order);
    DiscountResults discountResults = new DiscountResults(order.getBaseOrderAmount(), discountedItems);

    return new PromotionSummary(
            discountedItems.items(),
            discountResults.getTotalDiscountAmount(),
            discountResults.getFinalPaymentAmount(),
            Optional.ofNullable(discountResults.getBadge())
    );
}
```

seminss replied:
early-return! 할 수 있도록 습관을 들여야겠습니다. 📝
여러모로 좋은 의견 감사합니다 ☺️

Octoping925 commented on src/main/java/christmas/service/ChristmasDiscountCalculator.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    public DiscountedItems calculateDiscounts(VisitDate visitDate, Order order) {
+        List<DiscountAmount> discounts = new ArrayList<>();
+        discounts.add(discountDDay(visitDate));
+        discounts.add(discountWeekDay(visitDate, order));
+        discounts.add(discountWeekend(visitDate, order));
+        discounts.add(discountSpecialDay(visitDate));
+        discounts.add(discountGiveaway(order));
```
List.of를 활용해보면 좋을 거 같아요

Octoping925 commented on src/main/java/christmas/service/ChristmasDiscountCalculator.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    private boolean isWeekday(VisitDate visitDate) {
+        return eventSchedular.isWeekday(visitDate.getDate());
```
이건 VisitDate가 혼자서도 할 수 있는 일일 것 같아요

Octoping925 commented on src/main/java/christmas/model/DiscountResults.java:
```diff
@@ -0,0 +1,44 @@
+        return benefits.items().stream().mapToInt(DiscountAmount::amount).sum();
+    }
+
+    private int calculateTotalDiscountExcludingGiveaway(DiscountedItems benefits) {
+        int totalDiscount = 0;
+        for (DiscountAmount discount : benefits.items()) {
+            if (!discount.discountSettings().equals(GIVEAWAY_DISCOUNT)) {
+                totalDiscount += discount.amount();
+            }
+        }
+        return totalDiscount;
+    }
```
1. discountSetting은 enum이라 ==으로 비교할 수 있어요
2. 스트림을 써봤어도 이뻤을  거 같아요
```java
return benefits.items().stream()
        .filter(discount -> discount.discountSettings() != GIVEAWAY_DISCOUNT)
        .mapToInt(DiscountAmount::amount)
        .sum();
```

seminss replied:
뭔가 더럽다고 느껴졌던 코드가 스트림을 이용하니 훨씬 깔끔해지네요.. 👍👍

Octoping925 commented on src/main/java/christmas/view/input/validator/VisitDateValidator.java:
```diff
@@ -0,0 +1,28 @@
+
+import static christmas.exception.ValidationErrorMessage.*;
+
+public class VisitDateValidator {
```
스태틱 함수만 있는 유틸리티 클래스는 생성자를 private로 잠궈서 인스턴스화를 막아주면 좋스빈다

seminss replied:
중요한 디테일이네요. 감사합니다!

Octoping925 reviewed (commented):
고생하셨습니다

khyojun commented on src/main/java/christmas/config/Menu.java:
```diff
@@ -0,0 +1,54 @@
+        this.category = category;
+    }
+
+    public static Menu of(String menuName) {
```
Enum클래스들 내부에 of라는 메서드들이 있는데 해당하는 네이밍은 어떤 의미를 가지는지 궁금합니다!

seminss replied:
정적 펙토리 메서드 네이밍 컨벤션을 따랐습니다!

khyojun commented on src/main/java/christmas/exception/ValidationErrorMessage.java:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    public String getMessage() {
+        String errorMessage = "[ERROR] ";
```
이쪽 부분은 포맷팅을 이용해보시면 더 간결하게 작성할 수 있을거 같아요!

seminss replied:
이해가 잘 가지 않는데 혹시 예시를 들어주실 수 있을까요? 🥹

Octoping925 replied:
> 이쪽 부분은 포맷팅을 이용해보시면 더 간결하게 작성할 수 있을거 같아요!

"%s %s %s" 이런 스타일을 얘기하신걸까요

khyojun replied:
넵!

khyojun replied:
아니면 해당 부분은 한 번에 합친 형태로 묶어서 내보내도 좋을 거 같아요! String의 + 연산이 더 들어가서 보내지는 것보다는 StringBuilder방식으로 연산 후 보내도 될 거 같아요!

khyojun commented on src/main/java/christmas/exception/input/InvalidOrderException.java:
```diff
@@ -0,0 +1,7 @@
+package christmas.exception.input;
+
+public class InvalidOrderException extends IllegalArgumentException {
+    public InvalidOrderException(String message) {
```
커스텀 예외 인상 깊어요!

khyojun commented on src/main/java/christmas/model/policy/discount/ChristmasDiscountPolicy.java:
```diff
@@ -0,0 +1,33 @@
+
+import static christmas.model.policy.discount.DiscountSettings.*;
+
+public class ChristmasDiscountPolicy implements DiscountPolicy {
```
인터페이스 사용 인상 깊어요!

khyojun commented on src/main/java/christmas/view/input/parser/OrderParser.java:
```diff
@@ -0,0 +1,20 @@
+    public static List<SimpleEntry<String, Integer>> parseEachMenu(String userInput) {
+        return Arrays.stream(userInput.split(MENU_SEPARATOR.getValue()))
+                .map(item->{
+                    String[] parts = item.trim().split(MENU_CONNECTOR.getValue());
```
해당 부분에서 String[] 부분은 어떻게 나오게 된 것일까요? Stream 방식으로 사용되기 위해 쓰여진 부분일까요?

seminss replied:
주문메뉴-수량 을 파싱하면서 {주문메뉴, 수량}이 담겨 part[0]은 메뉴, part[1]은 수량을 갖고 있습니다. 항상 크기가 2이기 때문에 List보다는 배열을 사용했습니다.

khyojun commented on src/main/java/christmas/view/input/validator/OrderMenuValidator.java:
```diff
@@ -0,0 +1,29 @@
+        validateFormat(userInput);
+    }
+
+    private static void validateNotEmpty(String userInput) {
```
static으로 열리게 된 부분이 궁금하네요!

seminss replied:
OrderMenuValidator를 인스턴스로 선언하지 않고 static 메서드로 호출하게 한 이유는, 상태 없이 유틸적인 면만 가지기 때문이었습니다. validate 내부에서 사용되는 validateNotEmpty나 validateFormat도 이에 따라 private static으로 정의해줘야 했습니다!

khyojun replied:
하나의 static 으로 인하여서 어쩔 수 없이 private인데도 불구하고 static이 전파된 느낌이 들었던거 같아요!

https://tecoble.techcourse.co.kr/post/2020-07-16-static-method/

이 글 한 번 읽어보시고 나중에 어떤 방식이 좋을 지 판단해보시면 좋을 거 같아요!

khyojun commented on src/test/java/christmas/input/VisitDateValidatorTest.java:
```diff
@@ -0,0 +1,60 @@
+        validator.validate(userInput);
+    }
+
+    @DisplayName("음의 정수면 성공이다.")
```
해당 정수 테스트 세분화해서 나누신 거 인상 깊어요!

khyojun commented on src/test/java/christmas/input/VisitDateValidatorTest.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @DisplayName("문자가 들어오면 예외가 발생한다.")
+    @ValueSource(strings = {"&", "5&"})
```
해당 부분에 혹시나 하는 마음에 하이폰이나 콤마를 넣어봐도 좋았을 거 같아요!

seminss replied:
오 애플리케이션의 다른 부분에서 사용되는 문자를 넣어주는 게 확실히 더 좋은 테스트였겠네요. 조언 감사합니다!

khyojun commented on src/test/java/christmas/model/DiscountResultsTest.java:
```diff
@@ -0,0 +1,38 @@
+
+    @DisplayName("증정품을 받지 않으면 할인 후 예상 결제 금액은 총 주문 금액 - 총 혜택 금액이다.")
+    @Test
+    void finalPaymentAmountTest2() {
```
Displayname으로 이름이 다르게 나오긴 하지만 숫자로 1,2 보단 다른 네이밍을 하는게 좋아보여요!

khyojun commented on src/test/java/christmas/model/policy/calendar/ChristmasSpecialDayEvaluatorTest.java:
```diff
@@ -0,0 +1,50 @@
+        );
+    }
+
+    @DisplayName("3,10,17,24,25,31은 스페셜 데이다.")
```
이 부분은 저도 고민이긴 하지만 해당 방식으로 네이밍을 할 경우 이후 값이 변경 되어졌을때 테스트 케이스당 이름도 변경해야할 거 같아요!

그래서 저는

"스패셜 데이를 확인하기 위한 테스트!" 라는 형식으로 고치게 되었을때 이후 작업이 줄어들게 될 거 같은데 어떻게 생각하시나요!

seminss replied:
개인적으로는 현재 스페셜 데이가 언제인지 테스트케이스에서도 명시해 주는 것이 좋다고 생각해서 위와 같이 네이밍하기는 했습니다!
특히 저 같은 경우에는 스페셜 데이를 하드코딩 하지 않고, 아래와 같이 로직을 넣어 정의해주고 있어서 더더욱 명시해주는 게 중요하다고 생각했습니다.
```
private static Set<LocalDate> initializeSpecialDays() {
        Set<LocalDate> specialDays = new HashSet<>();
        //12월의 모든 일요일을 스페셜 데이에 추가
        LocalDate date = startDateOfDecember.with(TemporalAdjusters.firstInMonth(SUNDAY));
        while (date.getMonth() == DECEMBER) {
            specialDays.add(date);
            date = date.with(TemporalAdjusters.next(SUNDAY));
        }
        //크리스마스를 스페셜 데이에 추가
        specialDays.add(christmasDate);
        return specialDays;
    }
```
그러나 효준님 의견에도 어느정도 동의합니다!-!

khyojun reviewed (commented):
세민님 코드 보고 많이 배워가는거 같아요! ㅎㅎ 리뷰들도 좋은 리뷰들이 남아있어서 부족하지만 조금씩이라도 찾아서 리뷰 남겼어요!
4주간 고생하셨습니다!

develsvai commented on src/main/java/christmas/model/policy/calendar/ChristmasSpecialDayEvaluator.java:
```diff
@@ -0,0 +1,38 @@
+        LocalDate date = startDateOfDecember.with(TemporalAdjusters.firstInMonth(SUNDAY));
+        while (date.getMonth() == DECEMBER) {
+            specialDays.add(date);
+            date = date.with(TemporalAdjusters.next(SUNDAY));
```
이부분도 스트림으로 처리 하면 더 간결해 질수 있을 것 같아요!

`    specialDays.addAll(
        startDateOfDecember.datesUntil(startDateOfDecember.plusMonths(1).with(TemporalAdjusters.firstDayOfMonth()))
            .filter(date -> date.getDayOfWeek() == DayOfWeek.SUNDAY)
            .collect(Collectors.toSet())
    );`

seminss committed:
fix: 통과 안되던 단위 테스트 수정

BadgeTest: 할인 금액 양수 -> 음수
ChristmasDiscountCalculator: assertFalse -> assertTrue

seminss committed:
docs: 리펙토링 사항 추가

seminss committed:
refactor: inputView에서 받은 입력을 넘길 때 DTO로 넘긴다.

seminss committed:
refactor: Controller 구조 변경

Request에서 validate를 생성자 내부로 이동.
static으로 선언하지 않기 위함.

seminss committed:
refactor: 이벤트 날짜 Calender로 묶기

VisitDate 리펙토링, 테스트 코드 수정

seminss committed:
refactor: Order 리펙토링

Order 모델 리펙토링, 테스트 코드에 적용

seminss committed:
refactor: Badge의 null 반환을 제거, NONE 필드 생성

config 디렉토리를 model/constant로 수정

develsvai commented on src/main/java/christmas/model/DiscountedItems.java:
```diff
@@ -0,0 +1,8 @@
+package christmas.model;
+
+import christmas.model.vo.DiscountAmount;
+
+import java.util.List;
+
+public record DiscountedItems(List<DiscountAmount> items) {
+}
```
이 객채도 VO 인것 같다는 생각이 드는데 따로 Model로 빼두신 이유가 있을까요?

develsvai commented on src/main/java/christmas/service/ChristmasPromotionService.java:
```diff
@@ -0,0 +1,50 @@
+
+public class ChristmasPromotionService {
+
+    Order order;
+    VisitDate visitDate;
+    ChristmasDiscountCalculator discountCalculator = new ChristmasDiscountCalculator();
```
같은 계층에서 서로 의존성을 가지는 것은 피하는게 좋다고 알고 있어서 주입 받아서 사용했으면 더 좋았을거 같아요

develsvai reviewed (commented):
리뷰가 늦어 죄송합니다 ㅠ 코드를 읽을수록 부족한 점보다는 완성도 가 높다 느껴지는 코드였습니다. 수고 하셨습니다.!

seminss committed:
refactor: 파일 제거

seminss committed:
refactor: 할인 정책 별 클래스 정의

Promotion 인터페이스 정의 -> 할인과 이벤트로 기능 구현, 할인(Discount)를 4개의 할인 정책이 상속

seminss committed:
refactor: Event를 상속 받는 증정 이벤트 클래스 생성

다양한 이벤트가 생길 경우를 고려
증정 이벤트를 관리하는 Giveaway Enum 클래스

seminss committed:
refactor: 프로모션 일정을 관리하는 객체 PromotionPeriod 생성

뱃지 임계값 양수로 수정

seminss committed:
refactor: OrderItem을 OrderRequest 내부로 이동

seminss committed:
refactor: 할인 정책, 이벤트 종합해서 혜택 결과 만드는 객체 생성

seminss committed:
refactor: PromotionResponse 객체 생성, 빌더 생성

프로모션 이벤트 전체 결과(TotalPromotionOutcome)를 dto(PromotionResponse)로 변환해 outputview로 넘긴다.
변환하는 객체는 ResponseBuilder다.

seminss committed:
refactor: 생성자 주입으로 변환

seminss committed:
fix: 계산 버그 수

1. 총 혜택에 증정 이벤트 포함 안하던 부분 수정
2. 증정 이벤트 명을 제대로 출력 못하던 버그 수정
- GiveawayEvent가 증정품명(샴페인)을 출력하도록 오버라이딩 하고 있었다.
3. 최종 계산 금액에 혜택을 적용하지 않던 부분 수정

seminss committed:
Merge pull request #3 from seminss/refactor

Refactor
